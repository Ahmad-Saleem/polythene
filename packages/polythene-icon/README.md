# Icon

<a class="btn-demo" href="http://arthurclemens.github.io/Polythene-examples/index.html#/icon">Demo</a>

Displays an icon (image or [SVG](#svg)).



## Usage

### mSVG (recommended)

mSVG is an SVG string wrapped around `m.trust()` and stored as a Javascript file. This way the SVG can be imported just like any other module, without having to resort to bundler plugins.

Use option `msvg`:

~~~javascript
import m from "mithril";
import { icon } from "polythene-icon";
import gIconStars from "mmsvg/google/action/stars";

const myIcon = m(icon, {
  msvg: gIconStars
});
~~~

A large collection of msvg icons is available at [mmsvg](https://github.com/ArthurClemens/mmsvg).

You can create your own mSVG icons using this template:

~~~javascript
var m = require("mithril");
module.exports = m.trust("paste svg contents here");
~~~


### Using SVG files

To use SVG instead of an image, pass an [svg](#svg) option object:

~~~javascript
const myIcon = m(icon, {
  svg: {
    src: "img/arrow.svg"
  }
});
~~~


### Using image files

~~~javascript
import { icon } from "polythene-icon";

const myIcon = m(icon, {
  src: "img/arrow.png"
});
~~~



## Variations

* The size is set with option `type` (and of course with CSS, in conjuction with `class`).
* The color is set with the CSS `color` attribute of the parent element.



## Options

### Common component options

| **Parameter** |  **Mandatory** | **Type** | **Default** | **Description** |
| ------------- | -------------- | -------- | ----------- | --------------- |
| **element**   | optional | String | "div" | HTML element tag |
| **class**     | optional | String |       | Extra CSS class appended to `pe-icon` |
| **id**        | optional | String |       | HTML element id |
| **content**   | optional | Mithril element |  | Any content; replaces `vnode.children` and ignores `svg` and `msvg`  |
| **before**    | optional | Mithril element | | Extra content before main content; this content is placed left of subsequent elements with a lower stacking depth |
| **after**     | optional | Mithril element | | Extra content after main content; this content is placed right of preceding elements with a higher stacking depth |

### Icon options

| **Parameter** |  **Mandatory** | **Type** | **Default** | **Description** |
| ------------- | -------------- | -------- | ----------- | --------------- |
| **msvg**      | either `src` or `svg` or `msvg` must be passed | Mithril element |  | Mithril-ified SVG icon |
| **svg**       | either `src` or `svg` or `msvg` must be passed | Object |  | [svg](#svg) options object |
| **src**       | either `src` or `svg` or `msvg` must be passed | String |  | Icon URL (for `img` only) |
| **type**      | optional | String | "regular" | Sets the size: either "small" (16px), "regular" (24px), "medium" (32px), "large" (40px). Adds CSS class `pe-icon--small`, etcetera |
| **avatar**    | optional | Boolean | | Set to `true` to add class `pe-icon--avatar` which creates a round image |



## Composition

Icon is composed from:

* [SVG](#svg) (when using option `svg` or `msvg`)



## CSS classes

| **Element** | **Key**     |  **Class** |
| ----------- | ----------- | --------------- |
| Component   | component   | `pe-icon` |

| **State**         | **Key**     |  **Class** |
| ----------------- | ----------- | --------------- |
| Small             | small       | `pe-icon--small` |
| Regular           | regular     | `pe-icon--regular` |
| Medium            | medium      | `pe-icon--medium` |
| Large             | large       | `pe-icon--large` |
| Avatar round icon | avatar      | `pe-icon--avatar` |

