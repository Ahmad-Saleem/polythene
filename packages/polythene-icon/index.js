export { default } from "./src/icon";
export { default as classes } from "./src/classes";
export { default as vars } from "./src/theme/vars";
