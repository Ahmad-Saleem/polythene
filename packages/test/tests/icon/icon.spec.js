/* global describe, it, expect */
import m from "mithril";
import { tidy, defaultHtmlTidyOptions } from "mithril-jest";
import { tests } from "./tests.js";
import { runSnapshots } from "../../scripts/snapshots";
import { icon as component } from "polythene-icon";

runSnapshots(tests);

describe("Icon component", () => {
  it("no options", () => {
    const cmp = m(component);
    const html = tidy(cmp, {
      ...defaultHtmlTidyOptions,
      wrap: false,
      indent: false
    });
    expect(html).toContain("<div class=\"pe-icon pe-icon--regular  \"></div>");
  });
  it("option id", () => {
    const cmp = m(component, {id: "id-x"});
    const html = tidy(cmp);
    expect(html).toContain("id=\"id-x\"");
  });
  it("option class", () => {
    const cmp = m(component, {class: "class-x"});
    const html = tidy(cmp);
    expect(html).toContain("class-x\">");
  });
  it("option element", () => {
    const cmp = m(component, {element: "span"});
    const html = tidy(cmp);
    expect(html).toContain("<span ");
    expect(html).toContain("</span>");
  });
  it("option before", () => {
    const cmp = m(component, {before: m("span", "Before")});
    const html = tidy(cmp);
    expect(html).toContain("<span>Before</span>");
  });
  it("option after", () => {
    const cmp = m(component, {after: m("span", "After")});
    const html = tidy(cmp);
    expect(html).toContain("<span>After</span>");
  });
});
