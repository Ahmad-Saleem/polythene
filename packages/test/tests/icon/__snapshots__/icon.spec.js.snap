exports[`test Child node (svg children mmsvg) -- dark theme (color set with style option) 1`] = `
"<div style=\"color: rgb(255, 255, 255);\" class=
\"pe-icon pe-icon--regular\">
  <div class=\"pe-svg\">
    <svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" height=\"24\"
    viewbox=\"0 0 24 24\">
      <path d=
      \"M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm4.24 16L12 15.45 7.77 18l1.12-4.81-3.73-3.23 4.92-.42L12 5l1.92 4.53 4.92.42-3.73 3.23L16.23 18z\">
      </path>
    </svg>
  </div>
</div>
"
`;

exports[`test Child node (svg children mmsvg) 1`] = `
"<div class=\"pe-icon pe-icon--regular\">
  <div class=\"pe-svg\">
    <svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" height=\"24\"
    viewbox=\"0 0 24 24\">
      <path d=
      \"M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm4.24 16L12 15.45 7.77 18l1.12-4.81-3.73-3.23 4.92-.42L12 5l1.92 4.53 4.92.42-3.73 3.23L16.23 18z\">
      </path>
    </svg>
  </div>
</div>
"
`;

exports[`test Option: avatar (type large) 1`] = `
"<div class=\"pe-icon pe-icon--large pe-icon--avatar\">
  <img src=
  \"http://arthurclemens.github.io/assets/polythene/examples/avatar-1.png\">
</div>
"
`;

exports[`test Option: content (svg content mmsvg) 1`] = `
"<div class=\"pe-icon pe-icon--regular\">
  <div class=\"pe-svg\">
    <svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" height=\"24\"
    viewbox=\"0 0 24 24\">
      <path d=
      \"M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm4.24 16L12 15.45 7.77 18l1.12-4.81-3.73-3.23 4.92-.42L12 5l1.92 4.53 4.92.42-3.73 3.23L16.23 18z\">
      </path>
    </svg>
  </div>
</div>
"
`;

exports[`test Option: content (svg trusted content) 1`] = `
"<div class=\"pe-icon pe-icon--regular\">
  <div class=\"pe-svg\">
    <svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" height=\"24\"
    viewbox=\"0 0 24 24\">
      <path d=
      \"M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm4.24 16L12 15.45 7.77 18l1.12-4.81-3.73-3.23 4.92-.42L12 5l1.92 4.53 4.92.42-3.73 3.23L16.23 18z\">
      </path>
    </svg>
  </div>
</div>
"
`;

exports[`test Option: content 1`] = `
"<div class=\"pe-icon pe-icon--regular\">
  <svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" height=\"24\"
  viewbox=\"0 0 24 24\">
    <path d=
    \"M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm4.24 16L12 15.45 7.77 18l1.12-4.81-3.73-3.23 4.92-.42L12 5l1.92 4.53 4.92.42-3.73 3.23L16.23 18z\">
    </path>
  </svg>
</div>
"
`;

exports[`test Option: msvg (Zondicons) 1`] = `"<div class=\"pe-icon pe-icon--regular  \"><div class=\"pe-svg \"><svg width=\"20px\" height=\"20px\" viewBox=\"0 0 20 20\" version=\"1.1\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><title>wrench</title><desc>Created with Sketch.</desc><defs></defs><g id=\"Page-1\" stroke=\"none\" stroke-width=\"1\" fill=\"none\" fill-rule=\"evenodd\"><g id=\"wrench\" fill=\"#000000\"><path d=\"M9.42239525,7.09396812 C10.2387498,5.25806681 9.89409813,3.03105216 8.38844025,1.52539428 C6.93864576,0.0755997864 4.82002948,-0.297742297 3.02577331,0.40536803 L6.97422669,4.35382141 L4.14579956,7.18224853 L0.197346187,3.23379515 C-0.50576414,5.02805132 -0.132422056,7.1466676 1.31737244,8.59646209 C2.70997507,9.98906472 4.71967257,10.3884681 6.46624524,9.79467236 L16.5961941,19.9246212 L19.4246212,17.0961941 L9.42239525,7.09396812 Z\" id=\"Combined-Shape\"></path></g></g></svg></div></div>"`;

exports[`test Option: msvg 1`] = `
"<div class=\"pe-icon pe-icon--regular\">
  <div class=\"pe-svg\">
    <svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" height=\"24\"
    viewbox=\"0 0 24 24\">
      <path d=
      \"M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm4.24 16L12 15.45 7.77 18l1.12-4.81-3.73-3.23 4.92-.42L12 5l1.92 4.53 4.92.42-3.73 3.23L16.23 18z\">
      </path>
    </svg>
  </div>
</div>
"
`;

exports[`test Option: src (image file) 1`] = `
"<div class=\"pe-icon pe-icon--regular\">
  <img src=\"img/arrow-back.png\">
</div>
"
`;

exports[`test Option: src (svg file) 1`] = `
"<div class=\"pe-icon pe-icon--regular\">
  <img src=\"img/arrow-back.svg\">
</div>
"
`;

exports[`test Option: style 1`] = `
"<div style=\"color: rgb(239, 108, 0);\" class=
\"pe-icon pe-icon--regular\">
  <div class=\"pe-svg\">
    <svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" height=\"24\"
    viewbox=\"0 0 24 24\">
      <path d=
      \"M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm4.24 16L12 15.45 7.77 18l1.12-4.81-3.73-3.23 4.92-.42L12 5l1.92 4.53 4.92.42-3.73 3.23L16.23 18z\">
      </path>
    </svg>
  </div>
</div>
"
`;

exports[`test Option: type (large) 1`] = `
"<div class=\"pe-icon pe-icon--large\">
  <div class=\"pe-svg\">
    <svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" height=\"24\"
    viewbox=\"0 0 24 24\">
      <path d=
      \"M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm4.24 16L12 15.45 7.77 18l1.12-4.81-3.73-3.23 4.92-.42L12 5l1.92 4.53 4.92.42-3.73 3.23L16.23 18z\">
      </path>
    </svg>
  </div>
</div>
"
`;

exports[`test Option: type (medium) 1`] = `
"<div class=\"pe-icon pe-icon--medium\">
  <div class=\"pe-svg\">
    <svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" height=\"24\"
    viewbox=\"0 0 24 24\">
      <path d=
      \"M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm4.24 16L12 15.45 7.77 18l1.12-4.81-3.73-3.23 4.92-.42L12 5l1.92 4.53 4.92.42-3.73 3.23L16.23 18z\">
      </path>
    </svg>
  </div>
</div>
"
`;

exports[`test Option: type (regular) 1`] = `
"<div class=\"pe-icon pe-icon--regular\">
  <div class=\"pe-svg\">
    <svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" height=\"24\"
    viewbox=\"0 0 24 24\">
      <path d=
      \"M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm4.24 16L12 15.45 7.77 18l1.12-4.81-3.73-3.23 4.92-.42L12 5l1.92 4.53 4.92.42-3.73 3.23L16.23 18z\">
      </path>
    </svg>
  </div>
</div>
"
`;

exports[`test Option: type (small) 1`] = `
"<div class=\"pe-icon pe-icon--small\">
  <div class=\"pe-svg\">
    <svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" height=\"24\"
    viewbox=\"0 0 24 24\">
      <path d=
      \"M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm4.24 16L12 15.45 7.77 18l1.12-4.81-3.73-3.23 4.92-.42L12 5l1.92 4.53 4.92.42-3.73 3.23L16.23 18z\">
      </path>
    </svg>
  </div>
</div>
"
`;

exports[`test Themed (color and size) 1`] = `
"<div class=\"pe-icon pe-icon--regular tests-icon-themed-icon\">
  <div class=\"pe-svg\">
    <svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" height=\"24\"
    viewbox=\"0 0 24 24\">
      <path d=
      \"M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm4.24 16L12 15.45 7.77 18l1.12-4.81-3.73-3.23 4.92-.42L12 5l1.92 4.53 4.92.42-3.73 3.23L16.23 18z\">
      </path>
    </svg>
  </div>
</div>
"
`;
